src/libstore/s3-binary-cache-store.cc

```diff
@@ -184,6 +186,75 @@ S3Helper::FileTransferResult S3Helper::getObject(
     return res;
 }
 
+/* An adapter that can be used to convert a Sink to ostream or iostream.
+
+   Example:
+
+   SinkToBidirectionalDeviceAdapter deviceToSink(sink);
+   boost::iostreams::stream<SinkToBidirectionalDeviceAdapter> streamToSink(deviceToSink);
+
+ */
+struct SinkToBidirectionalDeviceAdapter {
+    typedef char char_type;
+    typedef boost::iostreams::bidirectional_device_tag category;
+    Sink & sink;
+    std::streamsize totalSize;
+
+    SinkToBidirectionalDeviceAdapter(Sink & sink) : sink(sink), totalSize(0) {}
+
+    std::streamsize read(char* data, std::streamsize dataSize)
+    {
+        // Intentionally reports no data read. This struct is only intended for sinks.
+        return 0;
+    }
+
+    std::streamsize write(const char* data, std::streamsize dataSize)
+    {
+        auto strViewData = std::string_view(data, dataSize);
+        this->sink(strViewData);
```

Comment:
We discovered that `sink()` uses boost coroutines underneath (in `sinkToSource`). It turns out that the coroutine system can throw a `forced_unwind` exception. Surprisingly, this exception is caught and swallowed by `std::ostream::write`. However, this exception is supposed to be caught by the boost coroutine system, which will handle it and report any user-code exceptions properly. So, now we only get an obfuscated error log that looks like this:
```
nix: include/boost/context/detail/exception.hpp:37: boost::context::detail::forced_unwind::~forced_unwind(): Assertion `caught' failed.
```
We believe we have a fix for this, but we still have to run it through its paces. Stay tuned.